src/libraries/System.Runtime/tests/System.Runtime.Extensions.Tests/System/Environment.StackTrace.cs

```diff
@@ -40,27 +46,27 @@ public void StackTraceTest()
             }
         }
 
-        [MethodImpl(MethodImplOptions.NoInlining)]
+        [MethodImpl(MethodImplOptions.NoOptimization | MethodImplOptions.NoInlining)]
```

Comment:
I also think we should explicitly have a test that demonstrates the behavior of a stack trace that occurs in optimized shared generic code where the shared generics context is no longer in scope. Even if it doesn't have exact type naming we still want to ensure it degrades gracefully rather than doing anything bad (like crashing).

Was there ever an agreed upon behavior for what text gets shown for the generic parameters in an optimized shared generic frame? I believe the text "System.__Canon" has already leaked out in a few other places such as the error handling text @hez2010 mentioned or it has shown up in VS debugger callstacks for many years when debugging optimized generic code:
<img width="443" alt="image" src="https://github.com/dotnet/runtime/assets/6243776/686abe6c-4f12-416f-b57d-9557edcfa15d">

I feel like the cat is already out of the bag in terms of the revealing that type name so using it here is aligned with precedent elsewhere. I'd also be fine reverting to use the type parameter name if you think its worth trying to get the cat back into the bag @jkotas?

Reply:
We do not leak `__Canon` through public APIs. We have always treated leaking `__Canon` through public APIs as bug.

I agree with https://github.com/dotnet/runtime/pull/96440#issuecomment-1999054175 .